### PR TITLE
RemovalsServletJakarta10 Recipe

### PR DESCRIPTION
**Pull request by ranuradh**
## What's changed?
This PR consists of a Jakarta EE 10 migration recipe - RemovalsServletJakarta10. Below is the screenshot of what the rule does:
image

image
image

## Anything in particular you'd like reviewers to focus on? -
**Update: This is fixed in the latest**
I was able to complete all the migrations except 
image

Here I need to be able to remove one argument using **org.openrewrite.java.DeleteMethodArgument** and then proceed to follow it up with another recipe. Unfortunately outside the test or within the test if I run , it recursively deletes all arguments and does not provide an option to stop after one cycle. In the recipe I have provided the details
```
  - org.openrewrite.java.DeleteMethodArgument:
      methodPattern: "B foo(int, int, int)"
      argumentIndex: 1
```
and the test has 
```
@Test
    void testY() {
        rewriteRun(
          //language=java
          java(b),
          java(
            "public class A {{ B.foo(0, 1, 2); }}",
            "public class A {{ B.foo(0, 2); }}"
          ));
    }
```
I would like the above result but without using cycles(1).expectedCyclesThatMakeChanges(1) I dont get the result and that wont work for me since I need to proceed and modify the result further.
Please let me know if you have further questions regarding the issue. Once I get past this I will be able to complete this recipe
## Anyone you would like to review specifically?
@timtebeek @joanvr 

## Have you considered any alternatives or workarounds?
Yes
### Checklist
- [X] I've added unit tests to cover both positive and negative cases
- [X] I've added the license header to any new files through `./gradlew licenseFormat`
- [X] I've used the IntelliJ IDEA auto-formatter on affected files

### CODE REVIEW

**Comment by timtebeek** [comment]
Not quite sure what's going on, but I can't seem to get the exception test cases to pass, which use a combination of `DeleteMethodArgument` and `ReorderMethodArguments`. @joanvr you recently patched DeleteMethodArgument; any idea what might be going on here?

**Comment by ranuradh** [comment]
Hi @timtebeek when I run the test with my last changes I see it passing - 
image

**Comment by timtebeek** [comment]
On my machine, and in CI, I'm getting 
```diff
@@ -5,8 +5,8 @@ 
 class Test {
     void doGet(HttpServletRequest req, HttpServletResponse res) throws Exception {
         jakarta.servlet.Servlet servlet ;
-        UnavailableException unavailableEx1 = new UnavailableException("x", 0);
-        UnavailableException unavailableEx2 = new UnavailableException("x", 0);
+        UnavailableException unavailableEx1 = new UnavailableException(0, "x");
+        UnavailableException unavailableEx2 = new UnavailableException(0, "x");
         UnavailableException unavailableEx3 = new UnavailableException( "x");
     }
 }
```
I'd like to see this work in CI before a merge; did you refresh your dependencies locally?

**Comment by timtebeek** [comment]
Also; at some point we might want to [update the README footer](https://github.com/openrewrite/rewrite-migrate-java#jakarta-ee-8-migration-recipes).

**Comment by ranuradh** [comment]
Hi @timtebeek will update the readme soon ..also I ran the same test in my branch ..
```
void testException() {
        rewriteRun(
          //language=java
          java("""
            package com.test;
             
            import java.io.IOException;
  
            import jakarta.servlet.ServletException;            
            import jakarta.servlet.SingleThreadModel;
            import jakarta.servlet.UnavailableException;
            import jakarta.servlet.http.HttpServlet;
            import jakarta.servlet.http.HttpServletRequest;
            import jakarta.servlet.http.HttpServletResponse;         
  
            public class Test {
            
                public void doGet(HttpServletRequest req, HttpServletResponse res) throws ServletException, IOException {
                
                 	   jakarta.servlet.Servlet servlet ;
                 	   UnavailableException unavailableEx2 = new UnavailableException("x",1);     
                 	   UnavailableException unavailableEx1 = new UnavailableException(0, null, "x");     
                 	   UnavailableException unavailableEx1 = new UnavailableException(0, servlet, "x");   
                       UnavailableException unavailableEx1 = new UnavailableException(servlet, "x");   
                 	    
                }
               
            }
            """, """
            package com.test;
             
            import java.io.IOException;
  
            import jakarta.servlet.ServletException;            
            import jakarta.servlet.SingleThreadModel;
            import jakarta.servlet.UnavailableException;
            import jakarta.servlet.http.HttpServlet;
            import jakarta.servlet.http.HttpServletRequest;
            import jakarta.servlet.http.HttpServletResponse;         
  
            public class Test {
            
                public void doGet(HttpServletRequest req, HttpServletResponse res) throws ServletException, IOException {
                
                 	   jakarta.servlet.Servlet servlet ;
                 	   UnavailableException unavailableEx2 = new UnavailableException("x");  
                 	   UnavailableException unavailableEx1 = new UnavailableException("x", 0);
                 	   UnavailableException unavailableEx1 = new UnavailableException("x", 0);    
                 	   UnavailableException unavailableEx1 = new UnavailableException("x");     
                }
               
            }
            """));
    }
```
and it ran without issues.. I am will pull all ur changes and run to see..

**Comment by ranuradh** [comment]
@timtebeek & @joanvr when I ran the debugger with just this 
```
java(
            """
              import jakarta.servlet.UnavailableException;
              import jakarta.servlet.http.HttpServletRequest;
              import jakarta.servlet.http.HttpServletResponse;
                
              class Test {
                  void doGet(HttpServletRequest req, HttpServletResponse res) throws Exception {
                      jakarta.servlet.Servlet servlet ;
                      UnavailableException unavailableEx1 = new UnavailableException(0, null, "x");
                    
                  }
              }
              """,
            """
              import jakarta.servlet.UnavailableException;
              import jakarta.servlet.http.HttpServletRequest;
              import jakarta.servlet.http.HttpServletResponse;
                
              class Test {
                  void doGet(HttpServletRequest req, HttpServletResponse res) throws Exception {
                      jakarta.servlet.Servlet servlet ;
                      UnavailableException unavailableEx1 = new UnavailableException("x", 0);
                     
                  }
              }
```
what I noticed is that after calling the DeleteMethodArgument class and getting upto deleting one argument, the test returns and fails instead of going to the next step of ReoderMethodArguments which inturn fails the test. I put breakpoints in ReorderMethodArguments but the code never reached there.

**Comment by ranuradh** [comment]
@joanvr and @timtebeek could you look at this failure? Thanks!

**Comment by timtebeek** [comment]
> @joanvr and @timtebeek could you look at this failure? Thanks!

Figured out why the reordering of arguments isn't yet working: the visitor we use there does not yet support constructors, [only method invocations](https://github.com/openrewrite/rewrite/blob/31d9eaa705c1ffc2ca471f3c40b36a49af586173/rewrite-java/src/main/java/org/openrewrite/java/ReorderMethodArguments.java#L123-L125). I'll see if I can add support there, such that we can use that here.

**Comment by timtebeek** [comment]
Now waiting for review and merge of
- https://github.com/openrewrite/rewrite/pull/3710

**Comment by joanvr** [comment]
I'm going to re-run CI on this PR when it finishes publishing the changes on `ReorderMethodArguments`

**Comment by joanvr** [comment]
Now that the CI tests pass, what's the status of this PR? is it ready to review/merge?

File: src/main/java/org/openrewrite/java/migrate/jakarta/GetRealPathUpdate.java
Comment (timtebeek) [inline src/main/java/org/openrewrite/java/migrate/jakarta/GetRealPathUpdate.java:47]:
```suggestion
    public String getDescription() {
```

File: src/main/java/org/openrewrite/java/migrate/jakarta/GetRealPathUpdate.java
Comment (timtebeek) [inline src/main/java/org/openrewrite/java/migrate/jakarta/GetRealPathUpdate.java:52]:
```suggestion
    public TreeVisitor<?, ExecutionContext> getVisitor() {
```

File: src/main/java/org/openrewrite/java/migrate/jakarta/GetRealPathUpdate.java
Comment (timtebeek) [inline src/main/java/org/openrewrite/java/migrate/jakarta/GetRealPathUpdate.java:63]:
```suggestion
```

File: src/main/java/org/openrewrite/java/migrate/jakarta/GetRealPathUpdate.java
Comment (ranuradh) [reply src/main/java/org/openrewrite/java/migrate/jakarta/GetRealPathUpdate.java]: updated in next commit

File: src/main/java/org/openrewrite/java/migrate/jakarta/GetRealPathUpdate.java
Comment (ranuradh) [reply src/main/java/org/openrewrite/java/migrate/jakarta/GetRealPathUpdate.java]: updated in next commit

File: src/main/java/org/openrewrite/java/migrate/jakarta/GetRealPathUpdate.java
Comment (ranuradh) [reply src/main/java/org/openrewrite/java/migrate/jakarta/GetRealPathUpdate.java]: updated in next commit

File: src/main/java/org/openrewrite/java/migrate/jakarta/GetRealPathUpdate.java
Comment (ranuradh) [reply src/main/java/org/openrewrite/java/migrate/jakarta/GetRealPathUpdate.java]: I renamed this and updated the display and description.
This class is to ```UpdateGetRealPath```
When we have the following : ```
jakarta.servlet.ServletRequest.getRealPath(String) and jakarta.servlet.ServletRequestWrapper.getRealPath(String)
```
we substitute it with req.getContext().getRealPath("") or reqWrapper.getContext().getRealPath(""). Hence I created a new class for the recipe.

File: src/main/java/org/openrewrite/java/migrate/jakarta/UpdateGetRealPath.java
Comment (timtebeek) [inline src/main/java/org/openrewrite/java/migrate/jakarta/UpdateGetRealPath.java:41]:
```suggestion
        return "Updates `getRealPath()` to call `getContext()` followed by `getRealPath()`";
```

**Comment by timtebeek** [comment]
Would be great if we could get these conflicts resolved before a more thorough review and merge.

**Review by timtebeek — APPROVED** [review]
The move of the faces recipes complicated the review a bit here, but I think I've resolved all the conflicts correctly now by comparing between `jakarta-ee-10.yml` and `jakarta-faces-4.yml`. Let me know if I missed any, but that greatly simplfied the review.

Thanks again!

**Pull request merged**
